BMM.UI.Android/Application/NewMediaPlayer/Playback/HttpSourceFactory.cs

```diff
@@ -7,22 +11,44 @@ namespace BMM.UI.Droid.Application.NewMediaPlayer.Playback
     public class HttpSourceFactory : Object, IDataSourceFactory
     {
         private readonly DefaultHttpDataSourceFactory _httpFactory;
-        private readonly IList<KeyValuePair<string, string>> _headers;
+        private readonly IAccessTokenProvider _accessTokenProvider;
+        private readonly IMediaRequestHttpHeaders _mediaRequestHttpHeaders;
 
-        public HttpSourceFactory(DefaultHttpDataSourceFactory httpFactory, IList<KeyValuePair<string, string>> headers)
+        public HttpSourceFactory(
+            DefaultHttpDataSourceFactory httpFactory,
+            IMediaRequestHttpHeaders mediaRequestHttpHeaders,
+            IAccessTokenProvider accessTokenProvider)
         {
             _httpFactory = httpFactory;
-            _headers = headers;
+            _mediaRequestHttpHeaders = mediaRequestHttpHeaders;
+            _accessTokenProvider = accessTokenProvider;
         }
 
         public IDataSource CreateDataSource()
         {
-            var source = _httpFactory.CreateDataSource() as DefaultHttpDataSource;
-            foreach (var header in _headers)
+            if (!(_httpFactory.CreateDataSource() is DefaultHttpDataSource source))
+                return null;
+            
+            if (!_accessTokenProvider.CheckIsTokenValid(_accessTokenProvider.AccessToken))
+                _accessTokenProvider.GetAccessToken().FireAndForget();
+
+            // Filtering out IAuthorizationHeaderProvider, because we need to avoid
+            // waiting for refresh token operation here, as this is sync method.
+            var headersProviders = _mediaRequestHttpHeaders
```

Comment:
Have you seen the class HttpHeaderProviders?
Right now we have different headers for UnauthorizedRequests, AuthorizedRequests and MediaRequests. Is it better to create another set for AndroidSyncMediaRequests? And then implement a NonAsyncAuthorizationHeaderProvider. (alternative name PotentiallyExpiredAuthorizationHeaderProvider)

Reply:
I agree, I will do it that way.
But I have one more idea. Not to request token refresh only in situation, when it already expired, but also when it is "about to expire". So let's say some user starts player, we detect that the token expires in one hour. To avoid the situation, when we have expired token in media player, we can request to refresh it now (in the background), not wait for the result, and still use the old token for current track (as it is still valid for an hour). In the worst scenario (e.g conection issues), the token won't refresh. We can also do it when opening the application.